### PR TITLE
TC39 Observable support

### PR DESCRIPTION
**Pull request by josh**
💡 Idea to add a first class API to make it easier to work with TC39 Observable interfaces. While TC39's proposal is stalled, a few libraries already implement the `Subscription` interface. This means compatibility with [RxJS](https://github.com/ReactiveX/RxJS) and [Zen Observables](https://github.com/zenparsing/zen-observable).

Instead of:

```js
function handler(event) {
  // ...
}

observe(".js-foo", {
  add(el) {
    el.addEventListener(el, "click", handler)
  },
  remove(el) {
    el.removeEventListener(el, "click", handler)
  }
})
```

Working with an observable library, you could:

```js
observe(".js-foo", {
  subscribe(el) {
    return Rx.Observable.fromEvent(el, "click").subscribe(event => {
      // ...
    })
  }
})
```

Which would automatically handle the unsubscribe when the element is no longer matching.

##

CC: @dgraham @mislav @muan @keithamus @koddsson

### CODE REVIEW

File: lib/index.js.flow [hunk 1]
```diff
@@ -27,7 +27,8 @@ type Options<T: Element> = {|
   constructor?: Class<T>,
   initialize?: InitializerCallback<T>,
   add?: AddCallback<T>,
-  remove?: RemoveCallback<T>
+  remove?: RemoveCallback<T>,
+  subscribe?: SubscribeCallback<T>
```
Comment (josh) [inline lib/index.js.flow]: An alternative might be to reuse `initialize` and support `InitializerCallbacks | Subscription` as a valid return object.

**Comment by keithamus** [comment]
The is a great first step - but I think the major innovation of Observables is encoding state as an object to pass around. This could be acheived here with some partial application:

```js
observe('.js-foo', { ... }) // looks for `.add` or `.subscribe` in options bag

const jsFoos = observe('.js-foo') // returns an "observe" observerable object

jsFoos.subscribe(() => { /* do stuff with this object */ })
```

Having this partial application style allows us to pass around an observable to different functions, including higher-order-observables like so:

```js
const jsFoos = observe('.js-foo')
jsFoosParents = map(jsFoos, (foo) => foo.parent) // get the parent of each `.js-foo`
jsFoosParentForms = filter(jsFoosParents, (fooParent) => isForm(fooParent)) // only emits where `.js-foo`s parent is a form
```

**Comment by josh** [comment]
@keithamus I've sketched out some APIs in that form as well, but no seems to many any of our common application use cases better.

One of the big features `observe` provides providing a cleanup handler as too. But creating an event stream of `add` doesn't give you that hook. You need to--maybe--emit both change add/remove events which just feels clumsy to work with. Do you have any API sketches of for how this could work?

**Comment by keithamus** [comment]
I guess there's two options, one is to subscribe to both add and remove events in one Observable, and filter them out for the individual event types, the other is to push that behaviour into observe with some mechanism like arguments or sub functions...

```js
const jsFoos = observe('.js-foo')
const addedFoos = filter(jsFoos, event => event.type === 'added')
const removedFoos = filter(jsFoos, event => event.type === 'removed')

// Or have some extra functions to handle this behaviour for you
const addedFoos = observe.added('.js-foo')
const removedFoos = observe.removed('.js-foo')
```

**Pull request merged**